File: extra-code/stage2_code/cpp_action_tutorial/src/count_until_server.cpp
// ============================================================
// Action 通信示例:CountUntil 服务端(Action Server)
// 演示:Action 的三段式接口 goal(target_number, period)/ feedback(current_number)
// / result(reached_number)。Action 适合"执行时间长、要汇报进度、可中途取消"的任务
// 收到目标后每隔 period 秒数一个数,边数边发 feedback,数到 target_number 返回 result
// 运行:ros2 run cpp_action_tutorial count_until_server
// ============================================================

#include <chrono>
#include <memory>
#include <thread>

// 自定义 Action 接口的头文件(由 count_until_interfaces 包构建时生成)
#include "count_until_interfaces/action/count_until.hpp"
#include "rclcpp/rclcpp.hpp"
// Action 的 API 在独立的 rclcpp_action 库中,不在 rclcpp 里
#include "rclcpp_action/rclcpp_action.hpp"

using CountUntil = count_until_interfaces::action::CountUntil;
// GoalHandle(目标句柄)代表服务端正在处理的一个 goal,
// 通过它发 feedback、查询取消状态、宣布成功/失败
using GoalHandleCountUntil = rclcpp_action::ServerGoalHandle<CountUntil>;

class CountUntilServer : public rclcpp::Node
{
public:
  CountUntilServer() : Node("count_until_server")
  {
    // Action Server 需要注册一组(共 3 个)回调,分别响应 goal 生命周期的不同阶段:
    //   handleGoal     - 收到新 goal 时决定接受还是拒绝(_1 是 goal 的 UUID,_2 是 goal 内容)
    //   handleCancel   - 客户端请求取消时决定是否允许
    //   handleAccepted - goal 被接受后开始真正执行
    server_ = rclcpp_action::create_server<CountUntil>(
      this,
      "count_until",
      std::bind(&CountUntilServer::handleGoal, this, std::placeholders::_1, std::placeholders::_2),
      std::bind(&CountUntilServer::handleCancel, this, std::placeholders::_1),
      std::bind(&CountUntilServer::handleAccepted, this, std::placeholders::_1));
    RCLCPP_INFO(get_logger(), "Action Server ready on /count_until");
  }

private:
  // 目标审核回调:先校验参数合法性,不合法直接 REJECT,合法则接受并立即执行
  rclcpp_action::GoalResponse handleGoal(
    const rclcpp_action::GoalUUID &,
    std::shared_ptr<const CountUntil::Goal> goal)
  {
    if (goal->target_number <= 0) {
      RCLCPP_WARN(get_logger(), "Rejecting goal: target_number must be positive");
      return rclcpp_action::GoalResponse::REJECT;
    }
    if (goal->period <= 0.0f) {
      RCLCPP_WARN(get_logger(), "Rejecting goal: period must be positive");
      return rclcpp_action::GoalResponse::REJECT;
    }
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  // 取消审核回调:本示例无条件允许取消(实际项目中可根据任务状态拒绝)
  rclcpp_action::CancelResponse handleCancel(const std::shared_ptr<GoalHandleCountUntil>)
  {
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  // goal 被接受后:在新线程中执行耗时任务,避免阻塞 spin 的事件循环
  // (回调里长时间阻塞会导致节点无法响应其他消息)detach 表示线程独立运行不必等待
  void handleAccepted(const std::shared_ptr<GoalHandleCountUntil> goal_handle)
  {
    std::thread{std::bind(&CountUntilServer::execute, this, goal_handle)}.detach();
  }

  // 真正的任务执行:循环数数,每一步发 feedback,期间随时检查是否被取消
  void execute(const std::shared_ptr<GoalHandleCountUntil> goal_handle)
  {
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<CountUntil::Feedback>();
    auto result = std::make_shared<CountUntil::Result>();
    // 用 std::max 给周期兜底,最小 0.1 秒,防止过快的循环
    auto period = std::chrono::duration<double>(std::max(0.1f, goal->period));

    for (int64_t number = 1; number <= goal->target_number; ++number) {
      // 客户端可能中途请求取消:确认后调用 canceled() 结束 goal 并返回部分结果
      if (goal_handle->is_canceling()) {
        result->reached_number = number - 1;
        goal_handle->canceled(result);
        return;
      }
      // publish_feedback 把当前进度实时推送给客户端(Action 特有,Service 做不到)
      feedback->current_number = number;
      goal_handle->publish_feedback(feedback);
      std::this_thread::sleep_for(period);
    }

    // 任务完成:succeed() 把最终 result 发给客户端并标记 goal 成功
    result->reached_number = goal->target_number;
    goal_handle->succeed(result);
  }

  rclcpp_action::Server<CountUntil>::SharedPtr server_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  // spin 处理 goal 请求、取消请求等事件;耗时执行已放到独立线程,不会卡住事件循环
  rclcpp::spin(std::make_shared<CountUntilServer>());
  rclcpp::shutdown();
  return 0;
}

File: extra-code/stage2_code/py_action_tutorial/py_action_tutorial/count_until_server.py
# count_until_server.py —— Action 服务器入门示例（从 1 数到目标值）
# 演示 Action 服务端的三个回调：
#   goal_callback   决定接受还是拒绝 Goal（GoalResponse）
#   cancel_callback 决定是否允许取消（CancelResponse）
#   execute_callback 真正干活：循环计数、发布 Feedback、返回 Result
# 运行：ros2 run py_action_tutorial count_until_server
import time

import rclpy
from count_until_interfaces.action import CountUntil
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import ExternalShutdownException, MultiThreadedExecutor
from rclpy.node import Node


class CountUntilServer(Node):
    def __init__(self):
        super().__init__("count_until_server")
        # ReentrantCallbackGroup：允许同组内多个回调并行执行。
        # execute_callback 里有 time.sleep 会长时间占用线程，
        # 若不并行，取消请求等其他回调会被卡住无法响应。
        self.callback_group_ = ReentrantCallbackGroup()
        # 创建 Action 服务器：指定 Action 类型、名称和三个回调
        self.action_server_ = ActionServer(
            self,
            CountUntil,
            "count_until",
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            callback_group=self.callback_group_,
        )
        self.get_logger().info("Action Server ready on /count_until")

    def goal_callback(self, goal_request):
        # Goal 审核：先做参数校验，不合法的目标直接 REJECT，
        # 这样 execute_callback 就只需处理合法输入
        if goal_request.target_number <= 0 or goal_request.period <= 0.0:
            return GoalResponse.REJECT
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        # 收到取消请求时的答复：ACCEPT 表示"允许取消"，
        # 真正停止计数的逻辑在 execute_callback 里检查
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        # Goal 被接受后由这里执行，返回值就是发回客户端的 Result
        feedback = CountUntil.Feedback()
        result = CountUntil.Result()
        period = max(0.1, float(goal_handle.request.period))

        for number in range(1, goal_handle.request.target_number + 1):
            # 每轮都要检查是否被取消，被取消则标记 canceled 并
            # 返回当前进度作为 Result
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                result.reached_number = number - 1
                self.get_logger().info(
                    f"Goal canceled at {result.reached_number}")
                return result
            # publish_feedback 让客户端实时看到执行进度，
            # 这正是 Action 相比 Service 的核心优势
            feedback.current_number = number
            goal_handle.publish_feedback(feedback)
            time.sleep(period)

        # 计数完成：标记成功并返回最终 Result
        result.reached_number = goal_handle.request.target_number
        goal_handle.succeed()
        return result


def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServer()
    # MultiThreadedExecutor：多线程执行器，配合 Reentrant 回调组，
    # 让执行中的 Goal 不会阻塞取消请求等其他回调
    executor = MultiThreadedExecutor()
    executor.add_node(node)
    try:
        executor.spin()
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        executor.shutdown(timeout_sec=2.0)
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
